### DISCUSSION TITLE
Carla0.9.14 crash

### DISCUSSION THREAD

**Discussion by walkincloud** [carla-simulator/carla#6218] [Q&A]
I am trying to run pygame simulation with carla 0.9.14 on Ubuntu 20.04, but sometimes crashes when simulation ends and here is the docker logs

CommonUnixCrashHandler: Signal=11
Unhandled Exception: SIGSEGV: invalid attempt to read memory at address 0x00000000000000a3

[2023.02.14-11.45.26:314][427]LogCore: Fatal error!

0x00000000029289ca CarlaUE4-Linux-Shipping!boost::asio::detail::epoll_reactor::deregister_descriptor(int, boost::asio::detail::epoll_reactor::descriptor_state*&, bool) [/mnt/sda3/carla/Unreal/CarlaUE4/Plugins/Carla/CarlaDependencies/include/boost/asio/detail/impl/epoll_reactor.ipp:384]
0x0000000005ede28f CarlaUE4-Linux-Shipping!boost::asio::detail::reactive_socket_service_base::close(boost::asio::detail::reactive_socket_service_base::base_implementation_type&, boost::system::error_code&)(+0x3e)
0x0000000005f2f5b0 CarlaUE4-Linux-Shipping!carla::streaming::detail::tcp::ServerSession::CloseNow()(+0xaf)
0x0000000005f32a85 CarlaUE4-Linux-Shipping!boost::asio::detail::completion_handler<boost::asio::detail::binder0<carla::streaming::detail::tcp::ServerSession::Close()::$_2>, boost::asio::io_context::basic_executor_type<std::__1::allocator<void>, 0ul> >::do_complete(void*, boost::asio::detail::scheduler_operation*, boost::system::error_code const&, unsigned long)()
0x0000000005ed9ad2 CarlaUE4-Linux-Shipping!boost::asio::detail::strand_service::do_complete(void*, boost::asio::detail::scheduler_operation*, boost::system::error_code const&, unsigned long)(+0x61)
0x0000000002913334 CarlaUE4-Linux-Shipping!boost::asio::detail::scheduler::do_run_one(boost::asio::detail::conditionally_enabled_mutex::scoped_lock&, boost::asio::detail::scheduler_thread_info&, boost::system::error_code const&) [/mnt/sda3/carla/Unreal/CarlaUE4/Plugins/Carla/CarlaDependencies/include/boost/asio/detail/impl/scheduler.ipp:492]
0x0000000002912e91 CarlaUE4-Linux-Shipping!boost::asio::detail::scheduler::run(boost::system::error_code&) [/mnt/sda3/carla/Unreal/CarlaUE4/Plugins/Carla/CarlaDependencies/include/boost/asio/detail/impl/scheduler.ipp:210]
0x0000000002912ca1 CarlaUE4-Linux-Shipping!boost::asio::io_context::run() [/mnt/sda3/carla/Unreal/CarlaUE4/Plugins/Carla/CarlaDependencies/include/boost/asio/impl/io_context.ipp:63]
0x0000000002912c07 CarlaUE4-Linux-Shipping!void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, carla::ThreadPool::AsyncRun(unsigned long)::'lambda'()> >(void*) [/home/test/UnrealEngine_4.26/Engine/Source/ThirdParty/Linux/LibCxx/include/c++/v1/thread:284]
0x00007fb1690246db libpthread.so.0!UnknownFunction(0x76da)
0x00007fb16801d88f libc.so.6!clone(+0x3e)

[2023.02.14-11.45.26:315][427]LogExit: Executing StaticShutdownAfterError
Engine crash handling finished; re-raising signal 11 for the default handler. Good bye.

It seems that the problem also happens in some previous versions and any suggestions?

**Comment by walkincloud** [comment]
The reason of dump is found, here is the log in which "debug 2" is added at the beginning of CloseNow function and ses means session_id, str means stream_id followed by thread id

2023-02-16T04:40:52.703187357Z  [2023.02.16-04.40.52:702][370]LogCarla: CarlaServer destroy_actor 179
2023-02-16T04:40:52.703189207Z  [2023.02.16-04.40.52:702][370]LogCarla: UActorDispatcher::Destroying actor: 'sensor.camera.rgb' 7ab512c0
2023-02-16T04:40:52.703190849Z  [2023.02.16-04.40.52:702][370]LogCarla:             SceneCaptureCamera_2147472978
2023-02-16T04:40:52.703192624Z  INFO:  debug 2: ses 54740 str 102 140396171294464 
2023-02-16T04:40:52.703194251Z  INFO:  session 54740 : error sending data : Broken pipe 140395786987264 
2023-02-16T04:40:52.703195959Z  INFO:  debug 2: ses 54740 str 102 140395786987264 
2023-02-16T04:40:52.715920065Z  CommonUnixCrashHandler: Signal=11
2023-02-16T04:40:52.737908241Z  [2023.02.16-04.40.52:737][371]LogCore: === Critical error: ===
2023-02-16T04:40:52.737909620Z  Unhandled Exception: SIGSEGV: invalid attempt to write memory at address 0x0000000000000003

Session 54740 of stream 102 has been closed by thread 140396171294464, but another thread 140395786987264 still tries to send message through the session due to scheduler sequence.  The session was closed and suggested broken pipe, then CloseNow is called again to access the invalid file descriptor.

To solve the problem, I added a member to class ServerSession indicating close status.  Whenever CloseNow is called, the status is checked and the following procedure is performed only if it is false, and it should be turned true before the function exits so that the next time the function is called will take no effect.  There should be a better solution.

**Reply by SampsonM8** [reply]
Hi ! Can u  please post your changes ? I am very interested in it.
